hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/InventoryStructure.java

```diff
@@ -381,8 +392,23 @@ XBB fillCommon(X entity, XBB bld) {
 
         for (SegmentType st : SegmentType.values()) {
             if (st != SegmentType.rl && st.isAllowedInCanonical() && check.canExtendTo(st)) {
-                Stream<?> next = getChildren(parent, Entity.entityTypeFromSegmentType(st));
-                ret = Stream.concat(ret, next);
+                //streams are a pain..
+                //we need to ensure that the children are evaluated 1 by one - 1 that no 2 queries run
+                //in parallel... in here we do not know if we are loading directly from a backend store
+                //or from memory.
+                //We cannot guarantee the transactional behavior of a backend - even if it supports a transaction frame
+                //it still may choose to use multiple transactions during the lifetime of a frame.
+                //Streams are evaluated lazily and concat'ed stream calls close on the underlying streams (i.e. commit()
+                //of the txs that we might be using) during its close method. As such, if we just concat'ed our calls
+                //to get children, we might try to nest the transactions - which is not supported by Titan at least.
+                //To overcome this, eagerly evaluate each of our children and compose the resulting stream from the
+                //loaded results.
+                List<?> res;
+                try (Stream<?> next = getChildren(parent, Entity.entityTypeFromSegmentType(st))) {
+                    res = next.collect(Collectors.toList());
+                }
+
+                ret = Stream.concat(ret, res.stream());
```

Comment:
The results are in essence iterated over twice and they all reside in memory, so yes.